rst_files/lake_model.rst

```diff
@@ -220,58 +220,26 @@ Here's the code:
 
 .. code-block:: julia
 
-    struct LakeModel{TF <: AbstractFloat}
-        λ::TF
-        α::TF
-        b::TF
-        d::TF
-        g::TF
-        A::Matrix{TF}
-        A_hat::Matrix{TF}
-    end
-
-    function LakeModel(;λ = 0.283,
-                        α = 0.013,
-                        b = 0.0124,
-                        d = 0.00822)
+    using Parameters, NLsolve
 
+    function LakeModel(;λ = 0.283, α = 0.013, b = 0.0124, d = 0.00822)
         g = b - d
-        A = [(1-λ) * (1-d) + b  (1-d) * α + b;
+        A = [(1-λ) * (1-d) + b  (1-d) * α + b; # carry out intermediate calculations
             (1-d) * λ          (1-d) * (1-α)]
         A_hat = A ./ (1 + g)
-
-        return LakeModel(λ, α, b, d, g, A, A_hat)
+        return (λ = λ, α = α, b = b, d = d, g = g, A = A, A_hat = A_hat) # return a NamedTuple
     end
 
-    function rate_steady_state(lm, tol = 1e-6)
-        x = 0.5 * ones(2)
-        error = tol + 1
-        while (error > tol)
-            new_x = lm.A_hat * x
-            error = maximum(abs, new_x - x)
-            x = new_x
-        end
-        return x
-    end
+    rate_steady_state(lm) = fixedpoint(x -> lm.A_hat * x, [0.5, 0.5], inplace = false).zero
 
     function simulate_stock_path(lm, X0, T)
-        X_path = zeros(eltype(X0), 2, T)
-        X = copy(X0)
-        for t in 1:T
-            X_path[:, t] = X
-            X = lm.A * X
-        end
-        return X_path
+        raw = [lm.A^t * X0 for t in 0:T-1]
+        reshaped = hcat(raw...)' # each row is one time period
```

Comment:
I believe `reduce(hcat, raw)'` avoids the splatting.

Reply:
We should avoid `reduce(hcat` since it is very unclear.  Why not just make a small funciton and broadcast for something so simple?

Reply:
Goes against the principle of using standard functions rather than defining your own, but how about
```
reduce((x, y) -> [x y], obj)
```

Reply:
That is impenetrable for someone just starting to program.

`reduce` is effectively banned in https://github.com/QuantEcon/lecture-source-jl/blob/master/style.md outside of parallel programming/etc.

It is too difficult for intro users to understand.  Look at the example there.  using `loops` with a little more code is easier for them to understand.

Reply:
Then it might might need a re-write of the structured used in the lecture.

Reply:
Sure.  Some simple for loops putting things in a matrix might make this very clear.  As long as we move away from a functional programming style for it, I am happy.

Reply:
We can definitely move away from the splatting, but I don't think that's a pressing concern. It seemed that @jlperla agreed, but let me know if I'm incorrect.

Reply:
No I hate the splatting.  Pretty much anytime the word `hcat` or `reduce` or `...` come up I am bound to hate it.

What is wrong with a for loop here?  Everybody can understand the loop equivalent to replace
```
        raw = [lm.A^t * X0 for t in 0:T-1]
        reshaped = hcat(raw...)' # each row is one time period
```
In fact, you can probably do it faster as well by going backwards in the iteration.  Something like
```
vals = # PREALLOCATE
vals[1,:] = X0
for t in 2:T
   vals[t, :] = lm.A * X
```
or something like that

Reply:
That works too, will do.